File: src/backtracking/Combinations.java
package backtracking;

import java.util.LinkedList;
import java.util.List;

import org.junit.Test;

/*
Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

For example,
If n = 4 and k = 2, a solution is:

[
  [2,4],
  [3,4],
  [2,3],
  [1,2],
  [1,3],
  [1,4],
]
 * */

public class Combinations 
{
    public List<List<Integer>> combine( int n, int k )
    {
    	List<List<Integer>> allCombs = new LinkedList<>();
    	List<Integer> oneComb = new LinkedList<>();
    	generateCombs( allCombs, oneComb, 0, n, k );
    	return allCombs;
    }
    
    private void generateCombs( List<List<Integer>> allCombs, List<Integer> oneComb, int start, int end, int k )
    {
    	if ( k == 0 )
    	{
    		allCombs.add( new LinkedList<>( oneComb ) );
    	}
    	
    	for ( int i = start; i <= end; i++ )
    	{
    		oneComb.add( i );
    		generateCombs( allCombs, oneComb, i + 1, end, k - 1 );
    		oneComb.remove( oneComb.size() - 1 );
    	}
    }
    
    @Test
    public void test()
    {
    	System.out.print( combine(3 ,2) );
    }
}

File: Python/Combinations/Combinations.py
# Definition for a binary tree node.
import unittest
from typing import List

"""
77. Combinations
Given two integers n and k, return all possible combinations of k numbers chosen from the range [1, n].

You may return the answer in any order.

Example 1:

Input: n = 4, k = 2
Output: [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]
Explanation: There are 4 choose 2 = 6 total combinations.
Note that combinations are unordered, i.e., [1,2] and [2,1] are considered to be the same combination.

Example 2:

Input: n = 1, k = 1
Output: [[1]]
Explanation: There is 1 choose 1 = 1 total combination.
"""

class TestCombinations(unittest.TestCase):


    def dfs(self, k: int, visited: List[bool], oneResult: List[int], allResults: List[List[int]]):
        if len(oneResult) > k:
            return
        elif len(oneResult) == k:
            allResults.append(oneResult.copy())
            return

        for i in range(len(oneResult), len(visited)):
            if not visited[i]:
                visited[i] = True
                oneResult.append(i + 1)
                self.dfs(k, visited, oneResult, allResults)
                oneResult.pop()
                visited[i] = False

        return

    def combine(self, n: int, k: int) -> List[List[int]]:
        allResults = []
        self.dfs(k, [False] * n, [], allResults)
        return allResults

    def test_combinations(self):

        abc = self.combine(4, 2)
        print(abc)

if __name__ == '__main__':
    unittest.main()